### PR TITLE
[PCP-21] Titan GC doesn’t affect online write

### PR DESCRIPTION
**Pull request by tabokie**
tikv/tikv#5739

Summary:
- support blob index as merge base value
- add value type to merge input
- support merge operation changes value type (restricted set)

### CODE REVIEW

File: db/merge_helper.cc [hunk 1]
```diff
@@ -319,15 +330,14 @@ Status MergeHelper::MergeUntil(InternalIterator* iter,
     assert(merge_context_.GetNumOperands() >= 1);
     assert(merge_context_.GetNumOperands() == keys_.size());
     std::string merge_result;
-    s = TimedFullMerge(user_merge_operator_, orig_ikey.user_key, nullptr,
-                       merge_context_.GetOperands(), &merge_result, logger_,
-                       stats_, env_);
+    s = TimedFullMerge(user_merge_operator_, orig_ikey.user_key, orig_ikey.type,
```
Comment (yiwu-arbug) [inline db/merge_helper.cc]: should pass `kTypeValue` by default?

File: db/version_set.cc [hunk 2]
```diff
@@ -1935,10 +1951,18 @@ void Version::MultiGet(const ReadOptions& read_options, MultiGetRange* range,
...
+          merge_operator_, user_key, base_type, nullptr,
+          iter->merge_context.GetOperands(), str_value, info_log_,
+          db_statistics_, env_, nullptr /* result_operand */, true);
+      if (base_type == kTypeBlobIndex) {
+        ROCKS_LOG_ERROR(info_log_, "Encounter unexpected blob index.");
+        *status = Status::NotSupported(
+            "Encounter unexpected blob index. Please open DB with "
+            "rocksdb::blob_db::BlobDB instead.");
+        return;
+      }
```
Comment (yiwu-arbug) [inline db/version_set.cc]: ok, I didn't realize Titan doesn't support MultiGet (needs to make `is_blob` a vector of boolean for each key being query).

File: table/get_context.cc [hunk 3]
```diff
@@ -257,13 +257,17 @@ bool GetContext::SaveValue(const ParsedInternalKey& parsed_key,
...
-                merge_operator_, user_key_, nullptr,
+                merge_operator_, user_key_, type, nullptr,
                 merge_context_->GetOperands(), pinnable_val_->GetSelf(),
                 logger_, statistics_, env_);
             pinnable_val_->PinSelf();
             if (!merge_status.ok()) {
               state_ = kCorrupt;
             }
+            if (is_blob_index_ != nullptr) {
+              // @TODO(tabokie): error if not supported.
```
Comment (yiwu-arbug) [inline table/get_context.cc]: return the error?

File: db/db_iter.cc [hunk 4]
```diff
@@ -667,6 +667,8 @@ inline bool DBIter::FindNextUserEntryInternal(bool skipping, bool prefix_check)
 // POST: saved_value_ has the merged value for the user key
 //       iter_ points to the next entry (or invalid)
 bool DBIter::MergeValuesNewToOld() {
+  // @TODO(tabokie): this function asserts we have value to yield.
```
Comment (yiwu-arbug) [inline db/db_iter.cc]: what does this todo mean? Cleanup the comments.

Comment (tabokie) [reply db/db_iter.cc]: (This is the major setback for delete by merge)

File: table/get_context.cc [hunk 5]
```diff
@@ -226,23 +226,27 @@ bool GetContext::SaveValue(const ParsedInternalKey& parsed_key,
...
-                merge_operator_, user_key_, &value,
-                merge_context_->GetOperands(), pinnable_val_->GetSelf(),
+                merge_operator_, user_key_, type, &value,
+                merge_context_->GetOperands(), &type, pinnable_val_->GetSelf(),
                 logger_, statistics_, env_);
             pinnable_val_->PinSelf();
             if (!merge_status.ok()) {
               state_ = kCorrupt;
+            } else if (is_blob_index_ != nullptr) {
+              *is_blob_index_ = (type == kTypeBlobIndex);
```
Comment (Connor1996) [inline table/get_context.cc]: should update `state_` too ?

Comment (tabokie) [reply table/get_context.cc]: I think state_ is kFound here. kBlobIndex is an abnormal state.

**Comment by Connor1996** [comment]
Please fix failed tests

**Comment by yiwu-arbug** [comment]
Let's send this PR to upstream.

**Comment by tabokie** [comment]
@yiwu-arbug facebook/rocksdb#6447 PTAL

**Pull request merged**